File: src/backtracing/_39_combination_sum.rs
struct Solution;

impl Solution {
    pub fn combination_sum(mut candidates: Vec<i32>, target: i32) -> Vec<Vec<i32>> {
        candidates.sort();
        let mut ret = vec![];
        fn helper(candidates: &Vec<i32>, target: i32, cur: &mut Vec<i32>, index: i32, ret: &mut Vec<Vec<i32>>) {
            let sum:i32 = cur.iter().sum();
            if sum == target {
                ret.push(cur.clone());
                return;
            }
            for i in index..candidates.len() as i32 {
                let v = candidates[i as usize];
                if sum + v > target {
                    break;
                }
                cur.push(v);
                helper(candidates, target, cur, i, ret);
                cur.pop();
            }
        }
        helper(&candidates,target,&mut vec![],0,&mut ret);
        ret
    }
}

File: hot-100/39_combinationSum.go
package hot_100

import "sort"

var ans [][]int

func combinationSum(candidates []int, target int) [][]int {
	ans = [][]int{}
	sort.Ints(candidates)
	var dfs func(cur []int, index int)
	dfs = func(cur []int, index int) {
		sum := 0
		for i := range cur {
			sum += cur[i]
		}
		if sum == target {
			tmp := make([]int, len(cur))
			copy(tmp, cur)
			ans = append(ans, tmp)
			return
		}
		for i := index; i < len(candidates); i++ {
			if sum+candidates[i] > target {
				break
			}
			cur = append(cur, candidates[i])
			dfs(cur, i)
			cur = cur[:len(cur)-1]
		}
	}
	dfs([]int{}, 0)
	return ans
}
